lib/Service/RoomService.php

```diff
@@ -416,6 +427,83 @@ public function setCallRecording(Room $room, int $status = Room::RECORDING_NONE)
 		$this->dispatcher->dispatch(Room::EVENT_AFTER_SET_CALL_RECORDING, $event);
 	}
 
+	public function storeRecording(Room $room, string $owner, array $file, IRootFolder $rootFolder): void {
+		if (is_null($file) || !is_array($file)) {
+			throw new InvalidArgumentException('no_file');
+		}
+
+		if (
+			$file['error'] !== 0 ||
+			!is_uploaded_file($file['tmp_name']) ||
+			Filesystem::isFileBlacklisted($file['tmp_name'])
+		) {
+			throw new InvalidArgumentException('invalid_file');
+		}
+
+		$content = file_get_contents($file['tmp_name']);
+		unlink($file['tmp_name']);
+
+		$mimeType = $this->mimeTypeDetector->detectString($content);
+		$allowedMimeTypes = [
+			'video/mp4',
+			'video/mpeg',
+			'video/ogg',
+			'audio/mp3',
+			'audio/ogg',
+		];
+		if (!in_array($mimeType, $allowedMimeTypes)) {
+			throw new InvalidArgumentException('file_mimetype');
+		}
+
+		$recordFileName = escapeshellcmd($file['name']);
+		if (strlen($recordFileName) !== strlen($file['name'])) {
+			throw new InvalidArgumentException('file_name');
+		}
+
+		$extensionFromMime = pathinfo(str_replace('/', '.', $mimeType), PATHINFO_EXTENSION);
+		$extensionFromFileName = strtolower(pathinfo($recordFileName, PATHINFO_EXTENSION));
+		if ($extensionFromFileName !== $extensionFromMime) {
+			throw new InvalidArgumentException('file_extension');
+		}
+
+		try {
+			$this->participantService->getParticipant($room, $owner);
+		} catch (ParticipantNotFoundException $e) {
+			throw new InvalidArgumentException('owner_participant');
+		}
+
+		try {
+			$recordingFolder = $this->getRecordingFolder($rootFolder, $owner, $room->getToken());
+			$recordingFolder->newFile($recordFileName, $content);
+		} catch (NoUserException $e) {
+			throw new InvalidArgumentException('owner_invalid');
+		} catch (NotPermittedException $e) {
+			throw new InvalidArgumentException('owner_permission');
+		}
+	}
+
+	private function getRecordingFolder(IRootFolder $rootFolder, string $owner, string $token): Folder {
+		$attachmentFolder = $this->config->getAttachmentFolder($owner);
+
+		$userFolder = $rootFolder->getUserFolder($owner);
+		try {
+			$attachmentFolder = $userFolder->get($attachmentFolder);
+		} catch (NotFoundException $e) {
+			$attachmentFolder = $userFolder->newFolder($attachmentFolder);
+		}
+		try {
+			$recordingRootFolder = $attachmentFolder->get('Recording');
+		} catch (NotFoundException $e) {
+			$recordingRootFolder = $attachmentFolder->newFolder('Recording');
+		}
+		try {
+			$recordingFolder = $recordingRootFolder->get($token);
+		} catch (NotFoundException $e) {
+			$recordingFolder = $recordingRootFolder->newFolder($token);
+		}
+		return $recordingFolder;
```

Comment:
@nickvergessen I stored the recorded file on this place:
`\Talk\Recording\<RoomToken>\<filename>`
Is this place ok? I created the `Recording` folder to make more easy to identify the list of recording files. Do you have other suggestion of folder structure that could be more readdable by users? Maybe using the pattern `<RoomName>_<RoomToken>` to folder name could be more readdable.

If this structure is ok, is there any simpler way to create the necessary folders? Sounds a bit ugly to me this try catch structure.

Other point is about the hardcoded name `Recording`. Do you think that will be good to create an appsetting to make possible change this value?

Reply:
The current pattern is `<RoomToken>`. We can maintain this and discuss about this and other attachments in other issue.

If you prefer to use other pattern, reply here reopening this comments that I implement.